### PR TITLE
Dazzerix Changes

### PR DESCRIPTION
**Pull request by Dazzerix**
I hope this is how I correctly do this. Also I'm feeling I wasted my time with the readme I setup due to how detailed this page is.

### CODE REVIEW

**Comment by Vendare** [comment]
Please always target the development branch for merges master is meant as the stable branch from which releases are made.

File: script/common/actor.js [hunk 1]
```diff
@@ -45,7 +48,8 @@ export class DarkHeresyActor extends Actor {
     this.system.insanityBonus = Math.floor(this.insanity / 10);
     this.system.corruptionBonus = Math.floor(this.corruption / 10);
     this.psy.currentRating = this.psy.rating - this.psy.sustained;
-    this.initiative.bonus = this.characteristics[this.initiative.characteristic].bonus;
+    if (this.type == "starship") {this.initiative.bonus = this.shipdetection;}
```
Comment (Vendare) [inline script/common/actor.js:51]: this should be ===. the == doesn't check the content its purely coincidence if it works as expected

File: script/sheet/item.js [hunk 2]
```diff
@@ -4,13 +4,24 @@ export class DarkHeresyItemSheet extends ItemSheet {
...
+    const data = await super.getData();
+    data.enrichment = await this._handleEnrichment();
+    data.system = data.data.system;
+    return data
   }
+  //  return {
+  //    item: data.item, 
+ //     system: data.data.system
+  //  };
+//  }
```
Comment (Vendare) [inline script/sheet/item.js:17]: Again delete old code that has been commented out if the new solution works

File: script/common/util.js [hunk 3]
```diff
@@ -28,9 +28,13 @@ export default class DarkHeresyUtil {
...
     rollData.clip= weapon.clip;
     rollData.rateOfFire= rateOfFire;
-    rollData.damageFormula= weapon.damage + (isMelee && !weapon.damage.match(/SB/gi) ? "+SB" : "");
-    rollData.penetrationFormula= weapon.penetration;
-    rollData.weaponTraits= this.extractWeaponTraits(weapon.special);    
+    rollData.weaponTraits= this.extractWeaponTraits(weapon.special); 
+    if (rollData.weaponTraits.rabbit && rollData.weaponTraits.force) {rollData.damageFormula= weapon.damage + "+" +actor.psy.rating + (isMelee && !weapon.damage.match(/AgB/gi) ? "+AgB" : "");}
+    else if (rollData.weaponTraits.force) {rollData.damageFormula= weapon.damage + "+" +actor.psy.rating + (isMelee && !weapon.damage.match(/SB/gi) ? "+SB" : "");}
+    else if (rollData.weaponTraits.rabbit) {rollData.damageFormula= weapon.damage + (isMelee && !weapon.damage.match(/AgB/gi) ? "+AgB" : "");}
+    else {rollData.damageFormula= weapon.damage + (isMelee && !weapon.damage.match(/SB/gi) ? "+SB" : "");}
```
[the remark this answers is not in the document]
Comment (Dazzerix) [reply script/common/util.js:35]: Thank you, that works quite well.

File: script/common/util.js [hunk 4]
```diff
@@ -118,6 +127,28 @@ export default class DarkHeresyUtil {
...
     }
   }
-    
+   
+  static createShipWeaponRollData(actor, starshipWeapon) {
+    let characteristic = this.getWeaponCharacteristic(actor, starshipWeapon);
+    let isCannon = starshipWeapon.comptype === "macro";
+    let isLance = starshipWeapon.comptype === "lance";
+    let isNova = starshipWeapon.comptype === "nova";
+    let isTorpedo = starshipWeapon.comptype === "torpedo";
```
[the remark this answers is not in the document]
Comment (Dazzerix) [reply script/common/util.js:136]: I didn't think of that, thank you. That is more efficient.

File: script/common/roll.js [hunk 5]
```diff
@@ -165,7 +191,8 @@ async function _computeDamage(damageFormula, penetration, dos, isAiming, weaponT
     replaced: false,
     damageRender: await r.render()
   };
-
+  if(weaponTraits.inaccurate){}
+  else
```
[the remark this answers is not in the document]
Comment (Dazzerix) [reply script/common/roll.js:195]: So when the accurate trait is looked for it creates a false positive if inaccurate is present. This looks for inaccurate first and stops the accurate function.

File: script/common/roll.js [hunk 6]
```diff
@@ -307,14 +337,19 @@ function _computeRateOfFire(rollData) {
     case "barrage":
       rollData.attackType.modifier = 0;
       rollData.attackType.hitMargin = 2;
-      rollData.maxAdditionalHit = rollData.rateOfFire.burst - 1;
+      if (rollData.weaponTraits.storm) 
+      {rollData.maxAdditionalHit = rollData.rateOfFire.burst *2 - 1;}
+      else
+      {rollData.maxAdditionalHit = rollData.rateOfFire.burst - 1;}
```
[the remark this answers is not in the document]
Comment (Dazzerix) [reply script/common/roll.js:343]: Thank you, I didn't know about this question mark function. This is much cleaner than if statements for everything.

File: script/common/roll.js [hunk 7]
```diff
@@ -343,6 +378,7 @@ function _computeRateOfFire(rollData) {
       rollData.attackType.hitMargin = 0;
       break;
   }
+  if (rollData.weaponTraits.twinLinked) {rollData.attackType.modifier = rollData.attackType.modifier + 20}
```
[the remark this answers is not in the document]
Comment (Dazzerix) [reply script/common/roll.js:381]: Thank you, changed where the code is implemented to compute target.

File: script/common/actor.js [hunk 8]
```diff
@@ -105,12 +109,9 @@ export class DarkHeresyActor extends Actor {
...
     for (let item of this.items) {
-      if (item.isTalent) {
-        this.experience.spentTalents += parseInt(item.cost, 10);
-      } else if (item.isPsychicPower) {
-        this.experience.spentPsychicPowers += parseInt(item.cost, 10);
-      }
-    }
+      if (item.isTalent) {this.experience.spentTalents += parseInt(item.cost, 10);}}
+    for (let item of this.items) {  
+      if (item.isPsychicPower) {this.experience.spentPsychicPowers += parseInt(item.cost, 10);}}
```
[the remark this answers is not in the document]
Comment (Dazzerix) [reply script/common/actor.js:114]: The fix for NPC experience was in the NPC file. I reverted this back to how it was before now.

**Review by Vendare — APPROVED** [review]

**Comment by TheGiddyLimit** [comment]
@Dazzerix I'm a little confused here--I can't find bits of this in Dark Heresy 2e. Are they from a different game? Or homebrew? 

Some examples:

- Rabbit Punch
- Nova Cannon
- Starships, generally

Where do these originate from?

**Comment by Dazzerix** [comment]
@TheGiddyLimit They are from the expanded RPG set. Everything about starships and their equipment is from rogue trader and its supplements. 
Rabbit punch is a talent found in the only war books that changes damage bonus from strength to agility. It's a ratling race only talent.

**Pull request merged**